packages/jsii/lib/transforms/deprecated-warnings-transformer.ts

```diff
@@ -0,0 +1,258 @@
+      ).filter(isClassOrInterfaceType);
+
+      for (const type of types) {
+        const node = bindings.getClassOrInterfaceRelatedNode(type)!;
+        this.index.set(fullyQualifiedName(this.typeChecker, node)!, node);
+      }
+
+      this.deprecatedTypes = Object.values(assembly.types)
+        .filter((typeInfo) => typeInfo.docs?.stability === Stability.Deprecated)
+        .filter(isClassOrInterfaceType)
+        .map((typeInfo) => bindings.getClassOrInterfaceRelatedNode(typeInfo)!)
+        .map((node) => this.typeChecker.getTypeAtLocation(node));
+    }
+  }
+}
+
+class DeprecatedWarningsTransformer {
+  private addWarning = false;
+  private warning?: Warning;
+
+  public constructor(
+    private readonly typeChecker: ts.TypeChecker,
+    private readonly context: ts.TransformationContext,
+    private readonly deprecatedTypes: ts.Type[],
+    private readonly index: Map<
+      string,
+      ts.ClassDeclaration | ts.InterfaceDeclaration | undefined
+    >,
+  ) {}
+
+  public transform<T extends ts.Node>(node: T): T {
+    return this.visitEachChild(node);
+  }
+
+  private visitEachChild<T extends ts.Node>(node: T): T {
+    return ts.visitEachChild(node, this.visitor.bind(this), this.context);
+  }
+
+  private visitor<T extends ts.Node>(node: T): ts.VisitResult<T> {
+    if (ts.isClassDeclaration(node)) {
+      const deprecatedType = this.findDeprecated([
+        this.typeChecker.getTypeAtLocation(node),
+      ]);
+      this.addWarning = deprecatedType != null;
+      if (this.addWarning) {
+        const deprecatedNode = this.index.get(
+          this.typeChecker.getFullyQualifiedName(deprecatedType!.symbol),
+        );
+        this.warning = getWarning(this.typeChecker, deprecatedNode!);
+      }
+    }
+
+    if (ts.isFunctionDeclaration(node)) {
+      this.addWarning = isDeprecated(node);
+      if (this.addWarning) {
+        this.warning = getWarning(this.typeChecker, node);
+      }
+    }
+
+    if (isMethodLikeDeclaration(node)) {
+      this.addWarning = this.addWarning || isDeprecated(node);
+      if (isDeprecated(node)) {
+        this.warning = getWarning(this.typeChecker, node);
+      }
+    }
+
+    if (ts.isBlock(node) && this.addWarning) {
+      const block = node as Block;
+      const warningStatement = this.createWarningStatement();
+      return ts.updateBlock(
+        node,
+        ts.createNodeArray([...warningStatement, ...block.statements]),
+      ) as any;
+    }
+
+    return this.visitEachChild(node);
+  }
+
+  private createWarningStatement() {
+    const warning = this.warning;
+    const message = `[WARNING] ${warning?.elementName} is deprecated.\n  ${warning?.message}\n  This API will be removed in the net major release.`;
+
+    // TODO There should be a simpler way...
+    return [
+      // TODO Instead of declaring the function everywhere it's used, should we create a single declaration somewhere and reuse it?
+      //  If so, where should this declaration go?
```

Comment:
Any ideas here are welcome.

Reply:
Is there any argument against adding a new file? Otherwise, I'm onboard with this idea.